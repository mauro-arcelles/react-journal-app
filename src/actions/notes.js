import { addDoc, collection, deleteDoc, doc, updateDoc } from 'firebase/firestore';
import Swal from 'sweetalert2';
import { db } from '../firebase/firebase-config';
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from '../helpers/loadNotes';
import { types } from '../types/types';

export const startNewNote = () => {
  return async (dispatch, getState) => {

    const { uid } = getState().auth;

    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    };

    const doc = await addDoc(collection(db, `${uid}/journal/notes/`), newNote);

    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote));

  };
};


export const activeNote = (id, note) => {
  return {
    type: types.notesActive,
    payload: {
      id,
      ...note
    }
  };
};

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const setNotes = (notes) => {
  return {
    type: types.notesLoad,
    payload: notes
  };

};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    const noteRef = doc(db, `${uid}/journal/notes/${note.id}`);
    await updateDoc(noteRef, noteToFirestore);

    dispatch(refreshNote(note.id, noteToFirestore));
    // dispatch(startLoadingNotes(uid));

    Swal.fire('Saved', 'Your note has been saved', 'success');

  };
};

export const refreshNote = (id, note) => {
  return {
    type: types.notesUpdated,
    payload: {
      id,
      note: {
        id,
        ...note
      }
    }
  };
};

export const startUploading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: 'Uploading...',
      text: 'Please wait',
      allowOutsideClick: false,
      didOpen: () => {
        Swal.showLoading();
      }
    });

    const fileUrl = await fileUpload(file);

    activeNote.url = fileUrl;
    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const uid = getState().auth.uid;

    const noteRef = doc(db, `${uid}/journal/notes/${id}`);
    await deleteDoc(noteRef);

    dispatch(deleteNote(id));
  };
};

export const deleteNote = (id) => {
  return {
    type: types.notesDelete,
    payload: id
  };
};

export const noteLogout = () => {
  return {
    type: types.notesLogoutCleaning
  };
};

export const addNewNote = (id, note) => {
  return {
    type: types.notesAddNew,
    payload: {
      id,
      ...note
    }
  };
};