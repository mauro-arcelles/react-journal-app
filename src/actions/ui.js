import { types } from '../types/types';

export const setError = (err) => {
  return {
    type: types.uiSetError,
    payload: err
  };
};

export const removeError = () => {
  return {
    type: types.uiRemoveError
  };
};

export const startLoading = () => {
  return {
    type: types.uiSetLoading
  };
};

export const removeLoading = () => {
  return {
    type: types.uiRemoveLoading
  };
};