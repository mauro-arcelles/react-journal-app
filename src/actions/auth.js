import { types } from '../types/types';
import { auth, googleAuthProvider } from '../firebase/firebase-config';
import { createUserWithEmailAndPassword, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from 'firebase/auth';
import { removeLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

// PARA RECORDAR:
// TODOS LAS ACCIONES ASINCRONAS TIENEN QUE SER UN METODO QUE RETORNA UN CALLBACK
// EN DONDE TENDRÁ EL dispatch Y el getState GRACIAS A redux-thunk


export const startLoginEmailPassword = (email, password) => {

  return (dispatch) => {
    dispatch(startLoading());

    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(removeLoading());
      })
      .catch(e => {
        dispatch(removeLoading());
        Swal.fire('Error', 'Username or password are incorrect', 'error');
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {

    signInWithPopup(auth, googleAuthProvider)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name });

        dispatch(login(user.uid, user.displayName));
      })
      .catch(e => {
        Swal.fire('Error', 'Username already exists', 'error');
      });
  };

};


export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName
  }
});


export const startLogout = () => {
  return async (dispatch) => {

    await signOut(auth);

    dispatch(logout());
    dispatch(noteLogout());

  };
};

export const logout = () => {
  return {
    type: types.logout
  };
};
