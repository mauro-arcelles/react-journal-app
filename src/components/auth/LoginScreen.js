import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';
import { useForm } from '../hooks/useForm';

export const LoginScreen = () => {

  const dispatch = useDispatch();
  const { loading } = useSelector(state => state.ui);

  const [formValues, handleInputChange] = useForm({
    email: 'nando@gmail.com',
    password: '123456'
  });

  const { email, password } = formValues;

  const handleLogin = (e) => {
    e.preventDefault();
    dispatch(startLoginEmailPassword(email, password));

  };

  const handleGoogleLogin = () => {
    dispatch(startGoogleLogin());
  };

  return (
    <>
      <h3 className='auth__title'>Login</h3>

      <form onSubmit={handleLogin} className='animate__animated animate__fadeIn animate__faster'>
        <input type='text' placeholder='Email' name='email' className='auth__input' autoComplete='off' value={email} onChange={handleInputChange} />
        <input type='password' placeholder='Password' name='password' className='auth__input' value={password} onChange={handleInputChange} />

        <button type='submit' className='btn btn-primary btn-block' disabled={loading}>
          {
            loading &&
            (<i className="fa-solid fa-spinner fa-spin"></i>)
          }
          &nbsp;INGRESAR
        </button>

        <div className='auth__social-networks'>
          <p>Login with social networks</p>

          <div className='google-btn' onClick={handleGoogleLogin}>
            <div className='google-icon-wrapper'>
              <img className='google-icon' src='https://upload.wikimedia.org/wikipedia/commons/thumb/2/2d/Google-favicon-2015.png/640px-Google-favicon-2015.png' alt='google button' width='100%' height='100' />
            </div>
            <p className='btn-text'>
              <b>Sign in with google</b>
            </p>
          </div>

        </div>
        <Link to='/auth/register' className='link'>
          Create new account
        </Link>

      </form>
    </>

  );
};
