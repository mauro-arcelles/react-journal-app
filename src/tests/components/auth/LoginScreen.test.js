import configureStore from 'redux-mock-store'; //ES6 modules
import thunk from 'redux-thunk';
import { mount } from 'enzyme/build';
import { Provider } from 'react-redux';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { MemoryRouter } from 'react-router-dom';
import { startGoogleLogin, startLoginEmailPassword } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => {
  return {
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn()
  };
});

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
  auth: {},
  ui: {
    loading: false,
    msgError: null
  }
};

let store = mockStore(initState);
store.dispatch = jest.fn();

describe('Pruebas en <LoginScreen />', () => {

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter>
        <LoginScreen />
      </MemoryRouter>
    </Provider>
  );

  beforeEach(() => {
    store = mockStore(initState);
    jest.clearAllMocks();
  });

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();

  });

  test('debe de disparar la accion de startGoogleLogin', () => {

    wrapper.find('.google-btn').prop('onClick')();

    expect(startGoogleLogin).toHaveBeenCalled();


  });

  test('debe de disparar el startLogin con los respectivos elementos', () => {

    wrapper.find('form').simulate('submit', {
      preventDefault: () => { }
    });

    expect(startLoginEmailPassword).toHaveBeenCalledWith('nando@gmail.com', '123456');

  });

});