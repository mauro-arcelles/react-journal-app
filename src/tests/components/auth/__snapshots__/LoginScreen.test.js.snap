// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginScreen /> debe de mostrarse correctamente 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Login
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
      value="nando@gmail.com"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value="123456"
    />
    <button
      className="btn btn-primary btn-block"
      disabled={false}
      type="submit"
    >
      Â INGRESAR
    </button>
    <div
      className="auth__social-networks"
    >
      <p>
        Login with social networks
      </p>
      <div
        className="google-btn"
        onClick={[Function]}
      >
        <div
          className="google-icon-wrapper"
        >
          <img
            alt="google button"
            className="google-icon"
            height="100"
            src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/2d/Google-favicon-2015.png/640px-Google-favicon-2015.png"
            width="100%"
          />
        </div>
        <p
          className="btn-text"
        >
          <b>
            Sign in with google
          </b>
        </p>
      </div>
    </div>
    <a
      className="link"
      href="/auth/register"
      onClick={[Function]}
    >
      Create new account
    </a>
  </form>,
]
`;
